proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,224 @@
+
+## Proposed solution
+
+Separate the responsibility of executing logic (Runnables/Callables/etc) from guiding when execution should take place (through retry policies, bulkheads, circuit breakers). In this way, failure handling strategies become configuration that can influence executions, and the execution API itself is just responsible for receiving some configuration and performing executions.
+
+By default, a failure handling strategy could assume, for example, that any exception is a failure. But in some cases, a `null` or negative return value could also be considered a failure. Users should be able to define this, and a user's definition of a failure is what should influence execution. (This all is what the Failsafe RetryPolicy's `retryOn`, `retryWhen`, `abortIf`, etc methods are all about - defining a failure).
+
+Standardise the Fallback, Bulkhead and CircuitBreaker APIs and provide implementations.
+
+* A Java API to provide RetryPolicy
+* A Java API to provide FallBack
+* A Java API to provide BulkHead
+* A Java API to provide CircuitBreaker
+
+## Detailed design
+
+
+### CDI-based approach 
+Use interceptor and annotation to specify the execution and policy configration.
+Define an interceptor binding to specify whether the execution is sync or asyn. Two interceptors should be provided by the implementation: AsyncExecutorInterceptor and SyncExecutorInterceptor
+```
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+@InterceptorBinding
+@Inherited
+public @interface Execution {
+
+    Type value() default Type.SYNC;
+
+    enum Type {
+        ASYNC, SYNC;
+    };
+}
+```
+#### RetryPolicy: A policy to define the retry criteria
+
+An annotation to specify the max retries, delays, maxDuration, Duration unit, jitter, retryOn, bakeOff, fallback etc.
+```
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+public @interface Retry {
+   
+    int maxRetries() default 3;
+
+    int delay() default 0;
+
+    TimeUnit delayUnit() default TimeUnit.SECONDS;
+
+    int maxDuration() default 20;
+    TimeUnit durationUnit() default TimeUnit.SECONDS;
+    int jitter() default 2;
+
+    int jitterFactor() default 2;
+
+    TimeUnit jitterDelayUnit() default TimeUnit.SECONDS;
+
+    int backOff() default 2;
+
+    int backOffFactor() default 2;
+
+    TimeUnit bakeOffUnit() default TimeUnit.SECONDS;
+
+    Class<? extends Throwable>[] retryOn() default { RuntimeException.class };
+
+    Class<? extends Throwable>[] aboartOn() default { RuntimeException.class };
+	String fallback() default "";
+}
+
+
+```
+#### CircuitBreaker: a rule to define when to close the circuit
+
+```
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Target({ ElementType.TYPE, ElementType.METHOD })
+public @interface CircuitBreaker {
```

Comment:
We need some way to define a fallback.  I've been experimenting with an `@Fallback` that takes a class that has the same interface as the class in use, but provides the degraded behavior in the case of an external failure.